tip-distributor/src/lib.rs

```diff
@@ -455,59 +455,70 @@ pub fn derive_tip_distribution_account_address(
     )
 }
 
-pub async fn send_transactions_with_retry(
+pub async fn sign_and_send_transactions_with_retries(
+    signer: &Keypair,
     rpc_client: &RpcClient,
-    transactions: &[Transaction],
-    max_send_duration: Duration,
-) -> usize {
-    let mut transactions_to_send: HashMap<Signature, Transaction> = transactions
-        .iter()
-        .map(|tx| (tx.signatures[0], tx.clone()))
-        .collect();
+    mut transactions: Vec<Transaction>,
+    max_retry_duration: Duration,
+) -> HashMap<Signature, Error> {
+    let mut errors = HashMap::default();
+    let mut blockhash = rpc_client
+        .get_latest_blockhash()
+        .await
+        .expect("fetch latest blockhash");
+    let mut signatures_to_transactions = transactions
+        .clone()
+        .into_iter()
+        .map(|tx| (tx.signatures[0], tx))
+        .collect::<HashMap<Signature, Transaction>>();
 
     let start = Instant::now();
-    while !transactions_to_send.is_empty() && start.elapsed() < max_send_duration {
-        info!("sending {} transactions", transactions_to_send.len());
-
-        for (signature, tx) in &transactions_to_send {
-            match rpc_client.send_transaction(tx).await {
-                Ok(send_tx_sig) => {
-                    info!("send transaction: {:?}", send_tx_sig);
-                }
-                Err(e) => {
-                    error!("error sending transaction {:?} error: {:?}", signature, e);
-                }
-            }
+    while start.elapsed() < max_retry_duration && !signatures_to_transactions.is_empty() {
+        if start.elapsed() > Duration::from_secs(60) {
+            blockhash = rpc_client
+                .get_latest_blockhash()
+                .await
+                .expect("fetch latest blockhash");
         }
 
-        sleep(Duration::from_secs(10)).await;
+        transactions
+            .iter_mut()
+            .for_each(|tx| tx.sign(&[signer], blockhash));
 
-        let mut signatures_confirmed: Vec<Signature> = Vec::new();
-        for signature in transactions_to_send.keys() {
-            match rpc_client.confirm_transaction(signature).await {
-                Ok(true) => {
-                    info!("confirmed signature: {:?}", signature);
-                    signatures_confirmed.push(*signature);
-                }
-                Ok(false) => {
-                    info!("didn't confirmed signature: {:?}", signature);
-                }
-                Err(e) => {
-                    error!(
-                        "error confirming signature: {:?}, signature: {:?}",
-                        signature, e
-                    );
+        let futs = transactions.iter().map(|tx| async {
```

Comment:
lots of copying between this + signatures_to_transactions, should be able to just use signatures_to_transactions and avoid the juggling of two variables